src/java.base/share/classes/java/lang/ref/Reference.java

```diff
@@ -235,61 +223,16 @@ private void enqueueFromPending() {
         if (q != ReferenceQueue.NULL) q.enqueue(this);
     }
 
-    private static final Object processPendingLock = new Object();
-    private static boolean processPendingActive = false;
-
     private static void processPendingReferences() {
         // Only the singleton reference processing thread calls
         // waitForReferencePendingList() and getAndClearReferencePendingList().
-        // These are separate operations to avoid a race with other threads
-        // that are calling waitForReferenceProcessing().
         waitForReferencePendingList();
-        Reference<?> pendingList;
-        synchronized (processPendingLock) {
-            pendingList = getAndClearReferencePendingList();
-            processPendingActive = true;
-        }
+        Reference<?> pendingList = getAndClearReferencePendingList();
         while (pendingList != null) {
             Reference<?> ref = pendingList;
             pendingList = ref.discovered;
             ref.discovered = null;
-
-            if (ref instanceof Cleaner) {
-                ((Cleaner)ref).clean();
-                // Notify any waiters that progress has been made.
-                // This improves latency for nio.Bits waiters, which
-                // are the only important ones.
-                synchronized (processPendingLock) {
-                    processPendingLock.notifyAll();
-                }
-            } else {
-                ref.enqueueFromPending();
-            }
-        }
-        // Notify any waiters of completion of current round.
-        synchronized (processPendingLock) {
-            processPendingActive = false;
-            processPendingLock.notifyAll();
-        }
-    }
-
-    // Wait for progress in reference processing.
-    //
-    // Returns true after waiting (for notification from the reference
-    // processing thread) if either (1) the VM has any pending
-    // references, or (2) the reference processing thread is
-    // processing references. Otherwise, returns false immediately.
-    private static boolean waitForReferenceProcessing()
```

Comment:
Removing waitForReferenceProcessing is contrary to 
https://bugs.openjdk.org/browse/JDK-8305186
Reference.waitForReferenceProcessing should be more accessible
I've pointed to JDK-8305186 in a number of discussions about problems encountered with various tests.
So I'm doubtful about removing it.

Reply:
No problem, I will revert those removals. I did the removals proactively, thinking none of this infrastructure is needed after we migrate DBB, but since there are apparently plausible use cases for it, I'll leave it in.

Reply:
Done in new commit, thanks for taking a look.